### PR TITLE
Refresh polish

### PR DESCRIPTION
**Pull request by hmlanigan**
Fix detail of refresh.  They should mimic how snap refresh works:

- `juju refresh <charm>` upgrades to the latest revision of the charm in the channel which was deployed. 
- `juju refresh <charm> --channel <risk>` upgrades to the latest revision of the charm in the track which was deployed, with the risk given. 
- `juju refresh <charm> --channel <track>/<risk>` upgrades to the latest revision of the charm in the channel provided.

Part of the fix is to ensure that the channel has been normalized where we need.  It should be normalized in the db.  However the input channel cannot be normalized before the determination is made on which channel to refresh too.

Side effort was to add verbose messaging into the fresher for debug purposes.

## QA steps

We can test moving around to the different channels easily.  Testing a refresh in the current channel requires uploading a new revision to a specific channel.

Note for testing, there is a bug with the info api, where we get channels which have been closed.  This may lead to some odd results if you try to refresh to a channel which is closed.

```console 
#
# Setup for deploy/refresh
# 
$ export JUJU_DEV_FEATURE_FLAGS="charm-hub"
$ juju bootstrap localhost test
$ juju add-model test --config charm-hub-url="https://api.staging.snapcraft.io" 
# 
# Get charm craft status on the charm, to know the expected results of refresh 
# 
$ charmcraft status --name ubuntu-qa
Track    Channel    Version    Revision
latest   stable     3          3
         candidate  1.1        4
         beta       ↑          ↑
         edge       ↑          ↑
1.0      stable     -          -
         candidate  1          1
         beta       ↑          ↑
         edge       2          2
2.0      stable     -          -
         candidate  -          -
         beta       -          -
         edge       3          3
# 
# Deploy, starting with the latest/candidate channel, refresh to the stable channel.  In this case the latest/stable has revision of 3.
# 
$ juju deploy ubuntu-qa --channel candidate
Located charm "ubuntu-qa-4".
Deploying charm "ubuntu-qa-4".
$ juju refresh ubuntu-qa --channel stable
Added charm "ubuntu-qa-3" to the model.
# 
# Deploy, starting with the 1.0/edge channel, refresh to the candidate channel.  In this case the 1.0/candidate has revision of 1.
# 
$ juju deploy ubuntu-qa --channel 1.0/edge
Located charm "ubuntu-qa-2".
Deploying charm "ubuntu-qa-2".
$ juju refresh ubuntu-qa --channel candidate
Added charm "ubuntu-qa-1" to the model.
$ juju refresh ubuntu-qa --channel latest/stable
Added charm "ubuntu-qa-3" to the model.
```

### CODE REVIEW

File: apiserver/facades/client/application/deploy.go [hunk 1]
```diff
@@ -168,17 +168,19 @@ func stateDeviceConstraints(cons map[string]devices.Constraints) map[string]stat
...
-			Track:  c.Track,
-			Risk:   string(c.Risk),
-			Branch: c.Branch,
+			Track:  normalizedC.Track,
+			Risk:   string(normalizedC.Risk),
+			Branch: normalizedC.Branch,
 		}
 	}
-	return &state.CharmOrigin{
+	stateOrigin := &state.CharmOrigin{
```
Comment (SimonRichardson) [inline apiserver/facades/client/application/deploy.go:178]:
```suggestion
	return &state.CharmOrigin{
```

File: cmd/juju/application/refresher/refresher_test.go [hunk 2]
```diff
@@ -473,3 +434,27 @@ func (s *charmHubCharmRefresherSuite) TestRefreshWithOriginChannel(c *gc.C) {
...
+}
+
+func refresherConfigWithOrigin(curl *charm.URL, ref string) RefresherConfig {
+	rc := basicRefresherConfig(curl, ref)
+	rc.CharmOrigin = corecharm.Origin{
+		Source:  corecharm.CharmHub,
+		Channel: &corecharm.Channel{},
+	}
+	return rc
+}
```
Comment (SimonRichardson) [inline cmd/juju/application/refresher/refresher_test.go:453]: This makes the tests sooo much cleaner!

**Review by SimonRichardson — COMMENTED** [review]
Code looks great, still got to do the Q&A steps.

**Review by SimonRichardson — APPROVED** [review]
Q&A was good :+1:

**Comment by hmlanigan** [comment]
$$merge$$

**Comment by hmlanigan** [comment]
$$merge$$

**Pull request merged**